File: Contract/src/test/java/org/cubexmc/contract/config/LanguageParityTest.java
package org.cubexmc.contract.config;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.bukkit.configuration.file.YamlConfiguration;
import org.junit.jupiter.api.Test;

/**
 * Guards the plugin's globalization contract:
 * <ul>
 *   <li>every bundled locale defines exactly the same key set, so switching {@code language}
 *       never silently degrades a screen to raw keys;</li>
 *   <li>every {@code ui.*} key the Kotlin sources ask for actually exists;</li>
 *   <li>translations keep the same {@code <placeholder>} tokens, so no locale drops a value.</li>
 * </ul>
 */
class LanguageParityTest {

    @Test
    void everyContractStatusHasALabelInBothLocales() {
        for (String locale : LOCALES) {
            YamlConfiguration lang = load(locale);
            for (org.cubexmc.contract.model.ContractStatus status : org.cubexmc.contract.model.ContractStatus.values()) {
                assertTrue(lang.isString("status." + status.name().toLowerCase(java.util.Locale.ROOT)),
                    locale + " is missing status " + status);
            }
        }
    }

    private static final List<String> LOCALES = List.of("zh_CN", "en_US");
    private static final Pattern PLACEHOLDER = Pattern.compile("<([a-z][a-z0-9_-]*)>");
    /**
     * The literal immediately inside a {@code ui(...)} call. Anchored on the call rather than the
     * whole line, because a line can legitimately mention other dashed literals — a {@code terms.*}
     * key passed as a placeholder value, for example.
     */
    private static final Pattern UI_KEY = Pattern.compile("\\bui\\(\\s*\"([a-z][a-z0-9-]*)\"");
    private static final Pattern FUNDING_FAIL_KEY = Pattern.compile("\\bfail\\(\\s*\"([a-z][a-z0-9-]*)\"");

    /** The conditional form: {@code ui(if (cond) "a" else "b")}. */
    private static final Pattern UI_KEY_CONDITIONAL = Pattern.compile(
            "\\bui\\(\\s*if\\s*\\([^)]*\\)\\s*\"([a-z][a-z0-9-]*)\"\\s*else\\s*\"([a-z][a-z0-9-]*)\"");

    @Test
    void bundledLocalesDefineTheSameKeys() {
        YamlConfiguration base = load(LOCALES.get(0));
        Set<String> baseKeys = base.getKeys(true);

        for (String locale : LOCALES.subList(1, LOCALES.size())) {
            YamlConfiguration other = load(locale);
            Set<String> otherKeys = other.getKeys(true);

            Set<String> missing = new TreeSet<>(baseKeys);
            missing.removeAll(otherKeys);
            Set<String> extra = new TreeSet<>(otherKeys);
            extra.removeAll(baseKeys);

            assertTrue(missing.isEmpty(), locale + ".yml is missing keys: " + missing);
            assertTrue(extra.isEmpty(), locale + ".yml has keys no other locale defines: " + extra);
        }
    }

    @Test
    void translationsKeepTheSamePlaceholders() {
        YamlConfiguration base = load(LOCALES.get(0));

        for (String locale : LOCALES.subList(1, LOCALES.size())) {
            YamlConfiguration other = load(locale);
            for (String key : base.getKeys(true)) {
                String baseValue = base.getString(key);
                String otherValue = other.getString(key);
                if (baseValue == null || otherValue == null || isCommandSyntax(key)) {
                    continue;
                }
                assertEquals(
                        placeholders(baseValue),
                        placeholders(otherValue),
                        "placeholder mismatch for " + key + " in " + locale + ".yml");
            }
        }
    }

    @Test
    void everyUiKeyReferencedInSourcesIsDefined() throws Exception {
        YamlConfiguration base = load(LOCALES.get(0));
        Set<String> defined = base.getKeys(true).stream()
                .filter(key -> key.startsWith("ui."))
                .map(key -> key.substring("ui.".length()))
                .collect(Collectors.toCollection(LinkedHashSet::new));

        Set<String> missing = new TreeSet<>();
        Path sources = Path.of("src", "main", "kotlin");
        try (Stream<Path> walk = Files.walk(sources)) {
            List<Path> kotlinFiles = walk.filter(path -> path.toString().endsWith(".kt")).toList();
            for (Path file : kotlinFiles) {
                String source = Files.readString(file, StandardCharsets.UTF_8);
                Matcher simple = UI_KEY.matcher(source);
                while (simple.find()) {
                    record(defined, missing, simple.group(1), file);
                }
                Matcher conditional = UI_KEY_CONDITIONAL.matcher(source);
                if (file.getFileName().toString().equals("AllianceFundingService.kt")) {
                    Matcher failures = FUNDING_FAIL_KEY.matcher(source);
                    while (failures.find()) record(defined, missing, failures.group(1), file);
                }
                while (conditional.find()) {
                    record(defined, missing, conditional.group(1), file);
                    record(defined, missing, conditional.group(2), file);
                }
            }
        }

        assertTrue(missing.isEmpty(), "ui keys used in code but not defined in lang files: " + missing);
    }

    @Test
    void noKotlinSourceEmitsLegacyColourCodes() throws Exception {
        // GuiItems no longer translates colours — every string reaches the player exactly as the
        // i18n service rendered it. A stray "&#RRGGBB" literal in Kotlin would therefore print
        // verbatim instead of colouring, which is easy to miss by eye and impossible to miss here.
        Pattern legacyColour = Pattern.compile("\"[^\"]*&#[0-9A-Fa-f]{6}");
        List<String> offenders = new ArrayList<>();
        try (Stream<Path> walk = Files.walk(Path.of("src", "main", "kotlin"))) {
            for (Path file : walk.filter(path -> path.toString().endsWith(".kt")).toList()) {
                List<String> lines = Files.readAllLines(file, StandardCharsets.UTF_8);
                for (int index = 0; index < lines.size(); index++) {
                    String line = lines.get(index);
                    if (line.stripLeading().startsWith("*") || line.stripLeading().startsWith("//")) {
                        continue;
                    }
                    if (legacyColour.matcher(line).find()) {
                        offenders.add(file.getFileName() + ":" + (index + 1));
                    }
                }
            }
        }

        assertTrue(offenders.isEmpty(),
                "legacy &#RRGGBB colour literals must live in the language files, not in Kotlin: " + offenders);
    }

    private void record(Set<String> defined, Set<String> missing, String key, Path file) {
        if (!defined.contains(key)) {
            missing.add(key + "  (" + file.getFileName() + ")");
        }
    }

    /**
     * Usage lines and the help block spell out command syntax like {@code <reward|item> <days>};
     * those angle brackets are argument names shown to the player, not substitution placeholders,
     * and each locale names them in its own language.
     */
    private boolean isCommandSyntax(String key) {
        return key.startsWith("ui.usage-") || key.equals("messages.help") || key.equals("messages.list-footer");
    }

    private Set<String> placeholders(String value) {
        Set<String> found = new TreeSet<>();
        Matcher matcher = PLACEHOLDER.matcher(value);
        while (matcher.find()) {
            found.add(matcher.group(1));
        }
        // <prefix> is substituted by the i18n layer, not by the caller.
        found.remove("prefix");
        return found;
    }

    private YamlConfiguration load(String locale) {
        File file = new File("src/main/resources/lang/" + locale + ".yml");
        assertTrue(file.isFile(), "missing bundled locale: " + file);
        return YamlConfiguration.loadConfiguration(file);
    }
}

File: StateCharge/src/test/kotlin/org/cubexmc/statecharge/config/LanguageParityTest.kt
package org.cubexmc.statecharge.config

import org.bukkit.configuration.file.YamlConfiguration
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.TreeSet
import java.util.regex.Pattern
import java.util.stream.Stream

/**
 * 守卫插件的国际化契约(照抄 Contract 的 LanguageParityTest):
 * - 每份 bundle 语言文件定义完全相同的 key 集,切换语言不会静默降级成裸 key;
 * - 源码里 `ui("...")` / `message("...")` 引用的 key 都真实存在;
 * - 各语言保持相同的 `<placeholder>` token;
 * - Kotlin 里没有 legacy 颜色字面量(所有可见颜色都在语言文件)。
 */
class LanguageParityTest {

    @Test
    fun bundledLocalesDefineTheSameKeys() {
        val base = load(LOCALES[0])
        val baseKeys = base.getKeys(true)

        for (locale in LOCALES.subList(1, LOCALES.size)) {
            val otherKeys = load(locale).getKeys(true)
            val missing = TreeSet(baseKeys).apply { removeAll(otherKeys) }
            val extra = TreeSet(otherKeys).apply { removeAll(baseKeys) }

            assertTrue(missing.isEmpty(), "$locale.yml is missing keys: $missing")
            assertTrue(extra.isEmpty(), "$locale.yml has keys no other locale defines: $extra")
        }
    }

    @Test
    fun translationsKeepTheSamePlaceholders() {
        val base = load(LOCALES[0])

        for (locale in LOCALES.subList(1, LOCALES.size)) {
            val other = load(locale)
            for (key in base.getKeys(true)) {
                val baseValue = base.getString(key)
                val otherValue = other.getString(key)
                if (baseValue == null || otherValue == null || isCommandSyntax(key)) {
                    continue
                }
                assertEquals(
                    placeholders(baseValue),
                    placeholders(otherValue),
                    "placeholder mismatch for $key in $locale.yml",
                )
            }
        }
    }

    @Test
    fun everyUiKeyReferencedInSourcesIsDefined() {
        val base = load(LOCALES[0])
        val defined = base.getKeys(true)
            .filter { it.startsWith("ui.") }
            .map { it.substring("ui.".length) }
            .toMutableSet()

        val missing = TreeSet<String>()
        val uiPattern = Pattern.compile("\\bui\\(\\s*\"([a-z][a-z0-9-]*)\"")
        for (source in kotlinSources()) {
            val matcher = uiPattern.matcher(source.first)
            while (matcher.find()) {
                if (!defined.contains(matcher.group(1))) {
                    missing.add(matcher.group(1) + "  (" + source.second.fileName + ")")
                }
            }
        }

        assertTrue(missing.isEmpty(), "ui keys used in code but not defined in lang files: $missing")
    }

    @Test
    fun everyMessageKeyReferencedInSourcesIsDefined() {
        val base = load(LOCALES[0])
        val defined = base.getKeys(true)
            .filter { it.startsWith("messages.") }
            .map { it.substring("messages.".length) }
            .toMutableSet()

        val missing = TreeSet<String>()
        val messagePattern = Pattern.compile("\\bmessage\\(\\s*\"([a-z][a-z0-9-]*)\"")
        for (source in kotlinSources()) {
            val matcher = messagePattern.matcher(source.first)
            while (matcher.find()) {
                if (!defined.contains(matcher.group(1))) {
                    missing.add(matcher.group(1) + "  (" + source.second.fileName + ")")
                }
            }
        }

        assertTrue(missing.isEmpty(), "message keys used in code but not defined in lang files: $missing")
    }

    @Test
    fun noKotlinSourceEmitsLegacyColourCodes() {
        // 所有玩家可见字符串都经语言文件渲染;Kotlin 里遗留的 "&#RRGGBB" 会原样打给玩家。
        val legacyColour = Pattern.compile("\"[^\"]*&#[0-9A-Fa-f]{6}")
        val offenders = mutableListOf<String>()
        for (source in kotlinSources()) {
            val lines = source.first.lines()
            for ((index, line) in lines.withIndex()) {
                if (line.trimStart().startsWith("*") || line.trimStart().startsWith("//")) {
                    continue
                }
                if (legacyColour.matcher(line).find()) {
                    offenders.add(source.second.fileName.toString() + ":" + (index + 1))
                }
            }
        }

        assertTrue(offenders.isEmpty(), "legacy &#RRGGBB colour literals must live in the language files, not in Kotlin: $offenders")
    }

    /**
     * usage 行与帮助块拼写命令语法(如 `<state> [count]`),尖括号是给玩家看的参数名而非
     * 替换占位符,各语言用自己的语言命名它们。
     */
    private fun isCommandSyntax(key: String): Boolean =
        key.startsWith("ui.usage-") || key == "messages.help"

    private fun placeholders(value: String): Set<String> {
        val found = TreeSet<String>()
        val matcher = PLACEHOLDER.matcher(value)
        while (matcher.find()) {
            found.add(matcher.group(1))
        }
        // <prefix> 由 i18n 层替换,不来自调用方。
        found.remove("prefix")
        return found
    }

    private fun kotlinSources(): List<Pair<String, Path>> {
        val sources = Path.of("src", "main", "kotlin")
        val files = mutableListOf<Pair<String, Path>>()
        Files.walk(sources).use { walk: Stream<Path> ->
            for (path in walk.filter { it.toString().endsWith(".kt") }.toList()) {
                files.add(Files.readString(path, StandardCharsets.UTF_8) to path)
            }
        }
        return files
    }

    private fun load(locale: String): YamlConfiguration {
        val file = File("src/main/resources/lang/$locale.yml")
        assertTrue(file.isFile, "missing bundled locale: $file")
        return YamlConfiguration.loadConfiguration(file)
    }

    private companion object {
        val LOCALES = listOf("zh_CN", "en_US")
        val PLACEHOLDER = Pattern.compile("<([a-z][a-z0-9_-]*)>")
    }
}
